Changed keep alive to erlang:send_after/3 style. [JIRA: RIAK-1887]

lordnull opened the pull request:
This resolves an issue where the provider process could become blocked or
stuck for an extended period of time, only to receive many pointless
messages from the timer system.

First change is to up the keepalive from 1 second to 60. This should be
enough for the remote end to realize if the socket connection has died, but
not so much as to be spammy.

The second, more invasive change, is to use the erlang:send_after/3 reset
pattern rather than timer:send_interval/3. By using the erlang:send_after,
messages will not be sent if the process becomes stuck, as no code capable
of resetting the timer will run. Secondly, the return values of the
erlang:send_after and erlang:cancel_timer don't need to be pattern matched
to ensure correctness (since the former returns the reference itself, and
the latter returns an integer or false).

Resolves #679 (RIAK-1807) .

lordnull committed:
Changed keep alive to erlang:send_after/3 style.

This resolves an issue where the provider process could become blocked or
stuck for an extended period of time, only to receive many pointless
messages from the timer system.

First change is to up the keepalive from 1 second to 60. This should be
enough for the remote end to realize if the socket connection has died, but
not so much as to be spammy.

The second, more invasive change, is to use the erlang:send_after/3 reset
pattern rather than timer:send_interval/3. By using the erlang:send_after,
messages will not be sent if the process becomes stuck, as no code capable
of resetting the timer will run. Secondly, the return values of the
erlang:send_after and erlang:cancel_timer don't need to be pattern matched
to ensure correctness (since the former returns the reference itself, and
the latter returns an integer or false).

jonmeredith commented on src/riak_repl2_pg_block_provider.erl:
```diff
@@ -109,7 +109,8 @@ handle_cast(_Msg, State) ->
     Data = term_to_binary(stay_awake),
     Transport:send(Socket, Data),
-    {noreply, State};
+    State2 = keepalive_timer(State),
```
Could modify state to set timer ref undefined here and avoid the cancel_timer.

jonmeredith commented on src/riak_repl2_pg_block_provider.erl:
```diff
@@ -160,6 +160,11 @@ terminate(_Reason, #state{}) ->
-keepalive_timer() ->
-    timer:send_interval(?KEEPALIVE, keepalive).
+keepalive_timer(State = #state{keepalive_timer = undefined}) ->
+    NewRef = erlang:send_after(?KEEPALIVE, keepalive, self()),
```
Does this work?  Manpage says

```
 erlang:send_after(Time, Dest, Msg) -> TimerRef
```

jonmeredith replied:
Doesn't error as as Dest can be a registered named.

```
Eshell V5.10.3  (abort with ^G)
1> erlang:send_after(60000, keepalive, self()).    
#Ref<0.0.0.28>
2> erlang:send_after(60000, self(), keepalive).
#Ref<0.0.0.33>
```

jonmeredith commented:
-1 timer interval not set correctly.

jonmeredith commented:
So, thinking about it, I'm not even sure if we should cancel and restart the timer on receiving TCP traffic.

Divining history based on other support incidents we've had, I think this timeout was added to handle the case of remote node death.  As the block provider connects _to_ the block requester and just listens for requests, if the requestor side dies, the provider does notice until the TCP keepalive mechanism discovers that (2 hours, 2 days, who knows, kernel/tuning specific), so this periodically generates traffic from provider to requestor to trigger TCP timeout logic.

I think it's safe just to periodically send the stay_awake message every 60s or so.

lordnull committed:
Fixed send_after call, and removed un-needed calls.

Based on feedback, resetting the keep-alive timer on each socket message
may not be strictly needed. We just need some data every 60 seconds or so
to ensure a node is alive (rather than the tcp timeout which could be
much longer). An interval is still not used (reasons in previous commit
still stand), so instead it's just re-creating the timer on each keepalive.

jonmeredith commented on src/riak_repl2_pg_block_provider.erl:
```diff
@@ -109,7 +109,8 @@ handle_cast(_Msg, State) ->
     Data = term_to_binary(stay_awake),
     Transport:send(Socket, Data),
-    {noreply, State};
+    State2 = keepalive_timer(State#state{keepalive_timer = undefined}),
```
Can clean this up - no longer cancel timers, so no need to do undefined distinction.

jonmeredith commented on src/riak_repl2_pg_block_provider.erl:
```diff
@@ -160,6 +158,11 @@ terminate(_Reason, #state{}) ->
+    NewRef = erlang:send_after(?KEEPALIVE, self(), keepalive),
+    State#state{keepalive_timer = NewRef};
 
+keepalive_timer(State) ->
```
This path is never hit.

jonmeredith commented on src/riak_repl2_pg_block_provider.erl:
```diff
@@ -160,6 +158,11 @@ terminate(_Reason, #state{}) ->
 
-keepalive_timer() ->
-    timer:send_interval(?KEEPALIVE, keepalive).
+keepalive_timer(State = #state{keepalive_timer = undefined}) ->
```
No longer need pattern match.

jonmeredith commented:
Confirmed existence of problem on existing code - new interval timer added for every served proxy get request

Set up two single node clusters c1 -> c2, enabled and started pg (provider on c1, requester c2)

```
{_,Pid,_,_}] = supervisor:which_children(riak_repl2_pg_block_requester_sup).
[riak_repl2_pg_block_requester:proxy_get(Pid, <<"b">>,<<"k">>,[]) || X <- lists:seq(1,1000)].
```

Verified timer increase with `ets:i(timer_interval_tab)`.

Verified keepalive message being reset with redbug

```
(dev1@127.0.0.1)27> redbug:start("riak_repl2_pg_block_provider:handle_info", [{msgs, 600},{time,300000}]).
ok

09:58:29 <{riak_repl2_pg_block_provider,init,1}> {riak_repl2_pg_block_provider,
                                                  handle_info,
                                                  [keepalive,
                                                   {state,ranch_tcp,
                                                    #Port<0.28119>,undefined,
                                                    "c2",#Ref<0.0.0.183149>,
                                                    undefined,
                                                    {riak_client,
                                                     'dev1@127.0.0.1',
                                                     undefined},
                                                    #Ref<0.0.0.207114>,3004}]}
09:59:29 <{riak_repl2_pg_block_provider,init,1}> {riak_repl2_pg_block_provider,
                                                  handle_info,
                                                  [keepalive,
                                                   {state,ranch_tcp,
                                                    #Port<0.28119>,undefined,
                                                    "c2",#Ref<0.0.0.183149>,
                                                    undefined,
                                                    {riak_client,
                                                     'dev1@127.0.0.1',
                                                     undefined},
                                                    #Ref<0.0.0.211602>,3004}]}
```

Will retest then approve after killing the dead code.

lordnull committed:
Removed keepalive_timer from state.

We never cancel it, nor refer to the timer in the code, ever. One could
flood the process by manually sending 'keepalive' to it, but if someone is
doing that, you've got bigger problems.

This means that now the timer is like a send interval that only intervals
if the process is not stuck, and does not add a timer on proxy get
requests.

Basho-JIRA commented:
Jon M confirmed code merged to 2.0.6 on 6/25/15.  Dev is done. Now just waiting to go into additional releases.  

**_[posted via JIRA by Patricia Brewer]**_
